src/UI/Component/Drilldown/Factory.php

```diff
@@ -0,0 +1,87 @@
+<?php
+declare(strict_types=1);
+
+/* Copyright (c) 2019 Nils Haagen <nils.haagen@concepts-and-training.de> Extended GPL, see docs/LICENSE */
+
+namespace ILIAS\UI\Component\Drilldown;
+
+/**
+ * Drilldown Factory
+ */
+interface Factory
+{
+	/**
+	 * ---
+	 * description:
+	 *   purpose: >
+	 *     A Drilldown Menu offers a partial view on a larger set of hierarchically
+	 *     structured navigation possibilities.
+	 *     While the entries of a Drilldown Menu are actually organised in a tree-structure,
```

Comment:
organised -> organized

Reply:
Hi @Amstutz ,
thanks for looking over this.

You are right, menu/submenu are merely identical; however, I need the distinction in the renderer, for JS-calls and templates are different after all. In the component-classes, I opted for typehinting and refrained from extending from a common base-class. That would be minimal change, though, and I'll build it in, if you want.

As for the naming - I do not have an actually opinion on that, but your suggestion sounds reasonable. Maybe have a short glance on this topic in JF and I'll adopt?

However, I find it important that there are always items in a submenu. Otherwise, the user might drill down to an empty list. 
With async-loading, that might happen easier, since I cannot really know without looking ahead.
Also, this would imho open wormholes when starting out in mid-level, i.e., not root, of the drilldown. Since drilldown-entries are limited to further menus and buttons, I do not think that the overhead of constructing the entire structure is too big.

But maybe I'm missing a point here - will the drilldown become a potential substitute for tree-navigation, like, on mobile devices? If so, I completely agree, and async loading would be mandatory.

Kindly,
NIls

Reply:
Hi @nhaagen 

Thanks a lot for the prompt feedback:

> You are right, menu/submenu are merely identical; however, I need the distinction in the renderer, for JS-calls and templates are different after all. In the component-classes, I opted for typehinting and refrained from extending from a common base-class. That would be minimal change, though, and I'll build it in, if you wan

I do not have a strong opinion, let's ask if there are other opinions at the upcoming JF, if not, I'll leave this up to you.

> As for the naming - I do not have an actually opinion on that, but your suggestion sounds reasonable. Maybe have a short glance on this topic in JF and I'll adopt?

+1, here my vote would go for having them in the same family. However, if most others like to have it the proposed way, I'll not go on the barricades. 

> However, I find it important that there are always items in a submenu. Otherwise, the user might drill down to an empty list. With async-loading, that might happen easier, since I cannot really know without looking ahead.
 
To prevent this, one should test this before adding an async submenu as node on the layer, right? To me, this seems feasible (add async menu only if it those contain children, otherwise not). 

> Also, this would imho open wormholes when starting out in mid-level, i.e., not root, of the drilldown. Since drilldown-entries are limited to further menus and buttons, I do not think that the overhead of constructing the entire structure is too big.

I am afraid, I do not understand this. Can you not limit, what is accepted as async response from the server by the JS? The pointer back to the parent is already given at the point of time you make the async request for the new submenu, the way up is always given by the initial rendering, only the way down can be extended by async requests.... Maybe I am missing something.

> But maybe I'm missing a point here - will the drilldown become a potential substitute for tree-navigation, like, on mobile devices?

To be honest, I do not know yet. The provided description is vague here. I thought about asking to flesh that out a little more, but did not, since I think this is to early to make a decision here. We need some more experience with those components. However I would like to see the drilldown to be ready for such requests, yes.